File: leetcode/859/c/main.c
#include <stdbool.h>
#include <stdlib.h>
#include <string.h>
bool buddyStrings(char* s, char* goal) {
    int s_size    = strlen(s);
    int goal_size = strlen(goal);

    if (s_size != goal_size) {
        return false;
    }
    if (strcmp(s, goal) == 0) {
        int ch_cnt[26] = { 0 };
        for (int i = 0; i < s_size; i++) {
            int idx = s[i] - 'a';
            if (ch_cnt[idx] == 1) {
                return true;
            }
            ch_cnt[idx]++;
        }
    }

    int* diffs = (int*)calloc(s_size, sizeof(int));
    int diff_cnt = 0;
    for (int i = 0; i < s_size; i++) {
        if (s[i] != goal[i]) {
            diffs[diff_cnt++] = i;
            if (diff_cnt > 2) {
                goto clean;
            }
        }
    }

    if (diff_cnt == 2) {
        if (s[diffs[0]] == goal[diffs[1]] && s[diffs[1]] == goal[diffs[0]]) {
            return true;
        }
    }

clean:
    free(diffs);
    return false;
}

File: leetcode/859/rs/src/main.rs
struct Solution;
impl Solution {
    pub fn buddy_strings(s: String, goal: String) -> bool {
        if s.len() != goal.len() {
            return false;
        }
        if s == goal {
            let mut ch_cnt = [0; 26];

            for &byte in s.as_bytes() {
                let idx = (byte - b'a') as usize;
                if ch_cnt[idx] == 1 {
                    return true;
                }
                ch_cnt[idx] += 1;
            }
        }
        let mut diffs = vec![];

        for (i, (ch1, ch2)) in s.chars().zip(goal.chars()).enumerate() {
            if ch1 != ch2 {
                diffs.push(i);
                if diffs.len() > 2 {
                    return false;
                }
            }
        }

        if diffs.len() == 2 {
            let a1 = s.as_bytes()[diffs[0]];
            let b1 = s.as_bytes()[diffs[1]];

            let a2 = goal.as_bytes()[diffs[0]];
            let b2 = goal.as_bytes()[diffs[1]];

            if a1 == b2 && b1 == a2 {
                return true;
            }
        }

        false
    }
}
fn main() {
    println!("Hello, world!");
}
